File: packages/sdk-python/examples/quickstart.py
"""End-to-end ZettaPay Python SDK quickstart.

Walks the full happy path against a live API:

    1. Health probe.
    2. Register a merchant.
    3. Read it back, list with pagination, patch it.
    4. Submit a base64-encoded x402 payment (requires a real signed tx).
    5. Clean up.

Run against a local API (no auth):

    python examples/quickstart.py

Run against a deployed environment:

    ZETTAPAY_BASE_URL=https://api.zettapay.dev \\
    ZETTAPAY_API_KEY=zp_live_... \\
    python examples/quickstart.py

The payment step is skipped unless ``ZETTAPAY_SIGNED_TX_BASE64`` is set —
the SDK does not sign transactions; produce a base64-encoded signed Solana
transfer with your wallet/keypair tooling and pass it via env var.
"""

from __future__ import annotations

import os
import sys
from typing import Optional

from zettapay import (
    RetryPolicy,
    ZettaPayClient,
    ZettaPayError,
    is_code,
    is_status,
)


def main() -> int:
    base_url = os.environ.get("ZETTAPAY_BASE_URL", "http://localhost:3000")
    api_key: Optional[str] = os.environ.get("ZETTAPAY_API_KEY")
    signed_tx: Optional[str] = os.environ.get("ZETTAPAY_SIGNED_TX_BASE64")

    client = ZettaPayClient(
        base_url,
        api_key=api_key,
        timeout=10.0,
        retry=RetryPolicy.default(),
    )

    print(f"→ ZettaPay quickstart against {base_url}")

    health = client.health()
    print(f"  health: status={health.status} merchants={health.merchants} payments={health.payments}")

    merchant = client.register_merchant(
        name="Acme Coffee",
        wallet_pubkey="7Np41oeYqPefeNQEHSv1UDhYrehxin3NStpSyab9YVhT",
        usdc_ata="EhpbDdUDKv2Ah6yyhyqz7n9zUQqvmW1qzPKNaqgQ4kZK",
    )
    print(f"  registered merchant id={merchant.id} name={merchant.name!r}")

    fetched = client.get_merchant(merchant.id)
    assert fetched.id == merchant.id
    print(f"  fetched merchant id={fetched.id} createdAt={fetched.created_at}")

    listing = client.list_merchants(limit=5)
    print(f"  list_merchants: count={listing.count} returned={len(listing.items)}")

    patched = client.update_merchant(merchant.id, name="Acme Coffee — Downtown")
    print(f"  patched merchant name={patched.name!r}")

    if signed_tx:
        receipt = client.pay(signed_tx)
        print(
            f"  pay: accepted={receipt.accepted} payment_id={receipt.payment_id} "
            f"feePayer={receipt.fee_payer}"
        )
        record = client.get_payment(receipt.payment_id)
        print(f"  get_payment: id={record.id} signers={len(record.signers)}")
    else:
        print("  pay: skipped (set ZETTAPAY_SIGNED_TX_BASE64 to exercise /pay)")

    client.delete_merchant(merchant.id)
    print(f"  deleted merchant id={merchant.id}")

    try:
        client.get_merchant(merchant.id)
    except ZettaPayError as err:
        if is_code(err, "not_found") or is_status(err, 404):
            print(f"  confirmed deletion (404 not_found)")
        else:
            raise

    print("✓ done")
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except ZettaPayError as err:
        print(
            f"✗ ZettaPayError code={err.code!r} status={err.status_code} message={err.message!r}",
            file=sys.stderr,
        )
        sys.exit(1)

File: packages/sdk-rust/examples/quickstart.rs
//! End-to-end ZettaPay Rust SDK quickstart.
//!
//! Walks the full happy path against a live API:
//!
//! 1. Health probe.
//! 2. Register a merchant.
//! 3. Read it back, list with pagination, patch it.
//! 4. Submit a base64-encoded x402 payment (requires a real signed tx).
//! 5. Clean up.
//!
//! ## Run
//!
//! Local dev API (no auth):
//!
//! ```bash
//! cargo run --example quickstart
//! ```
//!
//! Deployed environment:
//!
//! ```bash
//! ZETTAPAY_BASE_URL=https://api.zettapay.dev \
//! ZETTAPAY_API_KEY=zp_live_... \
//! cargo run --example quickstart
//! ```
//!
//! The payment step is skipped unless `ZETTAPAY_SIGNED_TX_BASE64` is set —
//! the SDK does not sign transactions. Produce a base64-encoded signed
//! Solana USDC transfer with your wallet/keypair tooling and pass it via
//! env var.

use std::env;
use std::time::Duration;

use zettapay::{
    Client, Error, ListOptions, RegisterMerchantInput, RetryPolicy, UpdateMerchantInput,
};

#[tokio::main]
async fn main() -> Result<(), Error> {
    let base_url = env::var("ZETTAPAY_BASE_URL").unwrap_or_else(|_| "http://localhost:3000".into());
    let api_key = env::var("ZETTAPAY_API_KEY").ok();
    let signed_tx = env::var("ZETTAPAY_SIGNED_TX_BASE64").ok();

    let mut builder = Client::builder()
        .base_url(&base_url)
        .timeout(Duration::from_secs(10))
        .retry(RetryPolicy::default_policy());
    if let Some(key) = api_key {
        builder = builder.api_key(key);
    }
    let client = builder.build()?;

    println!("→ ZettaPay quickstart against {base_url}");

    let health = client.health().await?;
    println!(
        "  health: status={} merchants={} payments={}",
        health.status, health.merchants, health.payments
    );

    let merchant = client
        .register_merchant(RegisterMerchantInput {
            name: "Acme Coffee".into(),
            wallet_pubkey: "7Np41oeYqPefeNQEHSv1UDhYrehxin3NStpSyab9YVhT".into(),
            usdc_ata: "EhpbDdUDKv2Ah6yyhyqz7n9zUQqvmW1qzPKNaqgQ4kZK".into(),
        })
        .await?;
    println!(
        "  registered merchant id={} name={:?}",
        merchant.id, merchant.name
    );

    let fetched = client.get_merchant(merchant.id).await?;
    assert_eq!(fetched.id, merchant.id);
    println!(
        "  fetched merchant id={} createdAt={}",
        fetched.id, fetched.created_at
    );

    let listing = client.list_merchants(ListOptions::new().limit(5)).await?;
    println!(
        "  list_merchants: count={} returned={}",
        listing.count,
        listing.items.len()
    );

    let patched = client
        .update_merchant(
            merchant.id,
            UpdateMerchantInput {
                name: Some("Acme Coffee — Downtown".into()),
                ..Default::default()
            },
        )
        .await?;
    println!("  patched merchant name={:?}", patched.name);

    if let Some(tx) = signed_tx {
        let receipt = client.pay_base64(&tx).await?;
        println!(
            "  pay: accepted={} payment_id={} feePayer={}",
            receipt.accepted, receipt.payment_id, receipt.fee_payer
        );
        let record = client.get_payment(&receipt.payment_id).await?;
        println!(
            "  get_payment: id={} signers={}",
            record.id,
            record.signers.len()
        );
    } else {
        println!("  pay: skipped (set ZETTAPAY_SIGNED_TX_BASE64 to exercise /pay)");
    }

    client.delete_merchant(merchant.id).await?;
    println!("  deleted merchant id={}", merchant.id);

    match client.get_merchant(merchant.id).await {
        Err(err) if err.is_code("not_found") || err.is_status(404) => {
            println!("  confirmed deletion (404 not_found)");
        }
        Err(err) => return Err(err),
        Ok(_) => {
            return Err(Error {
                code: "assertion_failed".into(),
                message: "merchant still exists after delete".into(),
                status_code: None,
                details: None,
            });
        }
    }

    println!("✓ done");
    Ok(())
}
